Bug: copy() fails with CROSSSLOT in Redis cluster mode

### Redis version

6.2.7

### Redisson version

4.3.1

### Redisson configuration

Default config

### What is the Expected behavior?

`RMap.copy()` and `RMapCache.copy()` should work in cluster mode.

### What is the Actual behavior?

When running `testCopy` and new test `testCopyPreservesTTL`, they crash with error:
```
CROSSSLOT Keys in request don't hash to the same slot
```
Raw tests output:
```
[ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 4.305 s <<< FAILURE! -- in org.redisson.RedissonMapCacheTest
[ERROR] org.redisson.RedissonMapCacheTest.testCopy -- Time elapsed: 0.093 s <<< ERROR!
org.redisson.client.RedisException: CROSSSLOT Keys in request don't hash to the same slot. channel: [id: 0x34732b84, L:/10.79.72.142:54643 - R:10.50.69.15/10.50.69.15:30246] command: (EVALSHA, cached script: local res = 0;local newKeysIndex = #KEYS/2; for j = 1, newKeysIndex, 1 do if tonumber(ARGV[1]) >= 0 then if ARGV[2] == '1' then res = res + redis.call('copy', KEYS[j], KEYS[newKeysIndex + j], 'db', AR...), params: [803c956f2096dc9b38f7000347acbed4ff2ca586, 10, test, redisson__timeout__set:{test}, redisson__idle__set:{test}, redisson__map_cache__last_access__set:{test}, {test}:redisson_options, test2, redisson__timeout__set:{test2}, redisson__idle__set:{test2}, ...], promise: java.util.concurrent.CompletableFuture@3ea11d58[Not completed, 1 dependents]

[ERROR] org.redisson.RedissonMapCacheTest.testCopyPreservesTTL -- Time elapsed: 0.034 s <<< ERROR!
org.redisson.client.RedisException: CROSSSLOT Keys in request don't hash to the same slot. channel: [id: 0x34732b84, L:/10.79.72.142:54643 - R:10.50.69.15/10.50.69.15:30246] command: (EVALSHA, cached script: local res = 0;local newKeysIndex = #KEYS/2; for j = 1, newKeysIndex, 1 do if tonumber(ARGV[1]) >= 0 then if ARGV[2] == '1' then res = res + redis.call('copy', KEYS[j], KEYS[newKeysIndex + j], 'db', AR...), params: [803c956f2096dc9b38f7000347acbed4ff2ca586, 10, testCopySrc, redisson__timeout__set:{testCopySrc}, redisson__idle__set:{testCopySrc}, redisson__map_cache__last_access__set:{testCopySrc}, {testCopySrc}:redisson_options, testCopyDst, redisson__timeout__set:{testCopyDst}, redisson__idle__set:{testCopyDst}, ...], promise: java.util.concurrent.CompletableFuture@43bf254e[Not completed, 1 dependents]
```
### Root cause
`RMapCache` has 5 associated keys per instance, all colocated via hash tag `{name}`. `copy("src", "dst")` packs all keys into a single EVAL. If `{src}` and `{dst}` hash to different slots, the cluster rejects it with CROSSSLOT.

The buggy code in `RedissonObject.java`:
```java
protected final RFuture<Boolean> copyAsync(CommandAsyncExecutor commandExecutor,
        List<Object> keys, int database, boolean replace) {
    // ...
    return commandExecutor.evalWriteAsync((String) keys.get(0), StringCodec.INSTANCE,
        RedisCommands.EVAL_BOOLEAN,
        "for j = 1, newKeysIndex, 1 do " +
            "res = res + redis.call('copy', KEYS[j], KEYS[newKeysIndex + j]); " +
        "end;",
        keys,   // ← all source + destination keys in one EVAL → CROSSSLOT
        database, Boolean.compare(replace, false));
}
```

### Additional information

_No response_

--- a/redisson/src/main/java/org/redisson/RedissonObject.java
+++ b/redisson/src/main/java/org/redisson/RedissonObject.java
@@ -250,38 +250,35 @@ public abstract class RedissonObject implements RObject {
             String destKey = (String) keys.get(pairCount + i);
 
             result = result.thenCompose(prevRes -> {
-                RFuture<Long> pttlFuture = commandExecutor.readAsync(sourceKey, StringCodec.INSTANCE, RedisCommands.PTTL, sourceKey);
-                return pttlFuture.thenCompose(ttl -> {
-                    if (ttl == -2) {
+                String dumpScript = "local t = redis.call('pttl', KEYS[1]); "
+                                  + "if t == -2 then return nil end; "
+                                  + "return {t, redis.call('dump', KEYS[1])};";
+
+                RFuture<List<Object>> dumpFuture = commandExecutor.evalWriteAsync(sourceKey, ByteArrayCodec.INSTANCE,
+                        RedisCommands.EVAL_LIST, dumpScript, Arrays.asList(sourceKey));
+                return dumpFuture.thenCompose(dumpResult -> {
+                    if (dumpResult == null) {
                         return CompletableFuture.completedFuture(prevRes);
                     }
-                    RFuture<byte[]> dumpFuture = commandExecutor.readAsync(sourceKey, ByteArrayCodec.INSTANCE, RedisCommands.DUMP, sourceKey);
-                    return dumpFuture.thenCompose(dumpBytes -> {
-                        if (dumpBytes == null) {
-                            return CompletableFuture.completedFuture(prevRes);
-                        }
-                        CompletionStage<Void> restoreFuture;
-                        if (!replace) {
-                            RFuture<Boolean> existsFuture = commandExecutor.readAsync(destKey, StringCodec.INSTANCE, RedisCommands.EXISTS, destKey);
-                            restoreFuture = existsFuture.thenCompose(exists -> {
-                                if (exists) {
-                                    return CompletableFuture.completedFuture(null);
-                                }
-                                long ttlMs = 0;
-                                if (ttl >= 0) {
-                                    ttlMs = ttl;
-                                }
-                                return commandExecutor.writeAsync(destKey, StringCodec.INSTANCE, RedisCommands.RESTORE, destKey, ttlMs, dumpBytes);
-                            });
-                        } else {
-                            long ttlMs = 0;
-                            if (ttl >= 0) {
-                                ttlMs = ttl;
-                            }
-                            restoreFuture = commandExecutor.writeAsync(destKey, StringCodec.INSTANCE, RedisCommands.RESTORE, destKey, ttlMs, dumpBytes, "REPLACE");
-                        }
-                        return restoreFuture.thenApply(v -> true);
-                    });
+                    long ttl = ((Number) dumpResult.get(0)).longValue();
+                    byte[] dumpBytes = (byte[]) dumpResult.get(1);
+
+                    long ttlMs = 0;
+                    if (ttl >= 0) {
+                        ttlMs = ttl;
+                    }
+                    String restoreScript = "if ARGV[1] == '1' or redis.call('exists', KEYS[1]) == 0 then "
+                                          + "redis.call('restore', KEYS[1], tonumber(ARGV[2]), ARGV[3]); "
+                                          + "return 1; "
+                                          + "end; "
+                                          + "return 0;";
+                    String replaceFlag = "0";
+                    if (replace) {
+                        replaceFlag = "1";
+                    }
+                    return commandExecutor.evalWriteAsync(destKey, StringCodec.INSTANCE, RedisCommands.EVAL_BOOLEAN,
+                            restoreScript, Arrays.asList(destKey), replaceFlag, ttlMs, dumpBytes)
+                            .thenApply(v -> true);
                 });
             });
         }

--- a/redisson/src/main/java/org/redisson/RedissonObject.java
+++ b/redisson/src/main/java/org/redisson/RedissonObject.java
@@ -190,6 +190,13 @@ public abstract class RedissonObject implements RObject {
 
     protected final RFuture<Boolean> copyAsync(CommandAsyncExecutor commandExecutor, List<Object> keys,
                                                 int database, boolean replace) {
+        int pairCount = keys.size() / 2;
+        if (getServiceManager().getCfg().isClusterConfig()
+                && commandExecutor.getConnectionManager().calcSlot((String) keys.get(pairCount))
+                    != commandExecutor.getConnectionManager().calcSlot((String) keys.get(0))) {
+            return executeDumpRestoreCopy(commandExecutor, keys, replace);
+        }
+
         if (keys.size() == 2) {
             List<Object> args = new ArrayList<>();
             args.add(keys.get(0));
@@ -225,7 +232,37 @@ public abstract class RedissonObject implements RObject {
                     + "return math.min(res, 1); ",
                     keys,
                     database, Boolean.compare(replace, false));
+    }
 
+    private RFuture<Boolean> executeDumpRestoreCopy(CommandAsyncExecutor commandExecutor, List<Object> keys,
+                                                     boolean replace) {
+        int pairCount = keys.size() / 2;
+        CompletableFuture<Boolean> result = CompletableFuture.completedFuture(false);
+
+        for (int i = 0; i < pairCount; i++) {
+            String sourceKey = (String) keys.get(i);
+            String destKey = (String) keys.get(pairCount + i);
+
+            result = result.thenCompose(prevRes -> {
+                RFuture<Long> pttlFuture = commandExecutor.readAsync(sourceKey, StringCodec.INSTANCE, RedisCommands.PTTL, sourceKey);
+                return pttlFuture.thenCompose(ttl -> {
+                    if (ttl == -2) {
+                        return CompletableFuture.completedFuture(prevRes);
+                    }
+                    RFuture<byte[]> dumpFuture = commandExecutor.readAsync(sourceKey, ByteArrayCodec.INSTANCE, RedisCommands.DUMP, sourceKey);
+                    return dumpFuture.thenCompose(dumpBytes -> {
+                        long ttlMs = ttl >= 0 ? ttl : 0;
+                        Object[] args = replace
+                                ? new Object[]{destKey, ttlMs, dumpBytes, "REPLACE"}
+                                : new Object[]{destKey, ttlMs, dumpBytes};
+                        return commandExecutor.writeAsync(destKey, StringCodec.INSTANCE, RedisCommands.RESTORE, args)
+                                .thenApply(v -> true);
+                    });
+                });
+            });
+        }
+
+        return new CompletableFutureWrapper<>(result);
     }
 
     protected final RFuture<Void> renameAsync(CommandAsyncExecutor commandExecutor, List<Object> keys, Runnable runnable) {

--- a/redisson/src/main/java/org/redisson/RedissonObject.java
+++ b/redisson/src/main/java/org/redisson/RedissonObject.java
@@ -251,12 +251,27 @@ public abstract class RedissonObject implements RObject {
                     }
                     RFuture<byte[]> dumpFuture = commandExecutor.readAsync(sourceKey, ByteArrayCodec.INSTANCE, RedisCommands.DUMP, sourceKey);
                     return dumpFuture.thenCompose(dumpBytes -> {
-                        long ttlMs = ttl >= 0 ? ttl : 0;
-                        Object[] args = replace
-                                ? new Object[]{destKey, ttlMs, dumpBytes, "REPLACE"}
-                                : new Object[]{destKey, ttlMs, dumpBytes};
-                        return commandExecutor.writeAsync(destKey, StringCodec.INSTANCE, RedisCommands.RESTORE, args)
-                                .thenApply(v -> true);
+                        CompletionStage<Void> restoreFuture;
+                        if (!replace) {
+                            RFuture<Boolean> existsFuture = commandExecutor.readAsync(destKey, StringCodec.INSTANCE, RedisCommands.EXISTS, destKey);
+                            restoreFuture = existsFuture.thenCompose(exists -> {
+                                if (exists) {
+                                    return CompletableFuture.completedFuture(null);
+                                }
+                                long ttlMs = 0;
+                                if (ttl >= 0) {
+                                    ttlMs = ttl;
+                                }
+                                return commandExecutor.writeAsync(destKey, StringCodec.INSTANCE, RedisCommands.RESTORE, destKey, ttlMs, dumpBytes);
+                            });
+                        } else {
+                            long ttlMs = 0;
+                            if (ttl >= 0) {
+                                ttlMs = ttl;
+                            }
+                            restoreFuture = commandExecutor.writeAsync(destKey, StringCodec.INSTANCE, RedisCommands.RESTORE, destKey, ttlMs, dumpBytes, "REPLACE");
+                        }
+                        return restoreFuture.thenApply(v -> true);
                     });
                 });
             });

--- a/redisson/src/main/java/org/redisson/RedissonObject.java
+++ b/redisson/src/main/java/org/redisson/RedissonObject.java
@@ -194,6 +194,12 @@ public abstract class RedissonObject implements RObject {
         if (getServiceManager().getCfg().isClusterConfig()
                 && commandExecutor.getConnectionManager().calcSlot((String) keys.get(pairCount))
                     != commandExecutor.getConnectionManager().calcSlot((String) keys.get(0))) {
+            if (commandExecutor instanceof BatchService) {
+                CompletableFuture<Boolean> f = new CompletableFuture<>();
+                f.completeExceptionally(new IllegalStateException(
+                        "COPY across different slots is not supported in batch/transaction mode"));
+                return new CompletableFutureWrapper<>(f);
+            }
             return executeDumpRestoreCopy(commandExecutor, keys, replace);
         }
 
@@ -251,6 +257,9 @@ public abstract class RedissonObject implements RObject {
                     }
                     RFuture<byte[]> dumpFuture = commandExecutor.readAsync(sourceKey, ByteArrayCodec.INSTANCE, RedisCommands.DUMP, sourceKey);
                     return dumpFuture.thenCompose(dumpBytes -> {
+                        if (dumpBytes == null) {
+                            return CompletableFuture.completedFuture(prevRes);
+                        }
                         CompletionStage<Void> restoreFuture;
                         if (!replace) {
                             RFuture<Boolean> existsFuture = commandExecutor.readAsync(destKey, StringCodec.INSTANCE, RedisCommands.EXISTS, destKey);

--- a/redisson/src/test/java/org/redisson/RedissonMapCacheTest.java
+++ b/redisson/src/test/java/org/redisson/RedissonMapCacheTest.java
@@ -1673,23 +1673,25 @@ public class RedissonMapCacheTest extends BaseMapTest {
     }
 
     @Test
-    public void testCopyPreservesTTL() throws InterruptedException {
-        RMapCache<String, String> map = redisson.getMapCache("testCopySrc");
-        map.put("key1", "val1", 10, TimeUnit.SECONDS);
-        map.put("key2", "val2");
+    public void testCopyPreservesTTL() {
+        testInCluster(redisson -> {
+            RMapCache<String, String> map = redisson.getMapCache("testCopySrc");
+            map.put("key1", "val1", 10, TimeUnit.SECONDS);
+            map.put("key2", "val2");
 
-        map.copy("testCopyDst");
+            map.copy("testCopyDst");
 
-        RMapCache<String, String> mapCopy = redisson.getMapCache("testCopyDst");
-        assertThat(mapCopy.get("key1")).isEqualTo("val1");
-        assertThat(mapCopy.get("key2")).isEqualTo("val2");
+            RMapCache<String, String> mapCopy = redisson.getMapCache("testCopyDst");
+            assertThat(mapCopy.get("key1")).isEqualTo("val1");
+            assertThat(mapCopy.get("key2")).isEqualTo("val2");
 
-        long ttl = mapCopy.remainTimeToLive("key1");
-        assertThat(ttl).isBetween(5000L, 10000L);
-        assertThat(mapCopy.remainTimeToLive("key2")).isEqualTo(-1);
+            long ttl = mapCopy.remainTimeToLive("key1");
+            assertThat(ttl).isBetween(5000L, 10000L);
+            assertThat(mapCopy.remainTimeToLive("key2")).isEqualTo(-1);
 
-        map.destroy();
-        mapCopy.destroy();
+            map.destroy();
+            mapCopy.destroy();
+        });
     }
 }
 

--- a/redisson/src/test/java/org/redisson/RedissonMapCacheTest.java
+++ b/redisson/src/test/java/org/redisson/RedissonMapCacheTest.java
@@ -1671,5 +1671,25 @@ public class RedissonMapCacheTest extends BaseMapTest {
 
         redisson.shutdown();
     }
+
+    @Test
+    public void testCopyPreservesTTL() throws InterruptedException {
+        RMapCache<String, String> map = redisson.getMapCache("testCopySrc");
+        map.put("key1", "val1", 10, TimeUnit.SECONDS);
+        map.put("key2", "val2");
+
+        map.copy("testCopyDst");
+
+        RMapCache<String, String> mapCopy = redisson.getMapCache("testCopyDst");
+        assertThat(mapCopy.get("key1")).isEqualTo("val1");
+        assertThat(mapCopy.get("key2")).isEqualTo("val2");
+
+        long ttl = mapCopy.remainTimeToLive("key1");
+        assertThat(ttl).isBetween(5000L, 10000L);
+        assertThat(mapCopy.remainTimeToLive("key2")).isEqualTo(-1);
+
+        map.destroy();
+        mapCopy.destroy();
+    }
 }
 

